### DISCUSSION TITLE
Backporting clangd fixes to LLVM point releases

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#1183] [General]
A user question that I encounter frequently is: will this clangd bugfix appear in an LLVM point release (e.g. 14.0.4) or do I have to wait until the next major release (e.g. 15.0)? I'm not sure how to answer and I would appreciate some guidance.

Specific questions I have on this topic:

**1. Where is the timeline for LLVM point releases documented?** What should I keep an eye on to know that e.g. "the point release LLVM 14.0.y is coming up, and we have until `<this date>` to backport fixes to it"?

**2. What criteria should be used to determine if a fix is suitable for backport? Who can/should make this call?**

**3. What is the process for performing a backport?** Is it just a matter of pushing to the relevant branch, or does some sort of permission/approval need to be attained first?

**Comment by HighCommander4** [comment]
FWIW, I ask in part because of a particular clangd 14 regression, #1100, that seems to be affecting a lot of users and I think would make a good candidate for backport to the 14.x branch.

**Comment by HighCommander4** [comment]
I found some information that may be relevant:

 * Point releases get a GitHub milestone which can be seen at https://github.com/llvm/llvm-project/milestones. The milestone can be in the "open" state or the "closed" state, so presumably when a point release's milestone is in an open state, it's accepting backports targeted at that release. I still don't know what is the schedule for point releases, or whether / where they are announced, or e.g. if there's going to be a 14.0.7 (I see the most recent one is 14.0.6), but I suppose keeping an eye on that milestones page is a start.

 * The procedure for requesting a backport seems to be documented here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches
   * There is some additional information in the [discussion](https://discourse.llvm.org/t/rfc-new-automated-release-workflow-using-issues-and-pull-requests/59473) about the introduction of this process
   * [This page](https://www.llvm.org/docs/HowToReleaseLLVM.html#triaging-bug-reports-for-releases) describes more of the process from a release manager's point of view.

So, for a clangd backport, I guess it would make sense to file an issue specifically for the backport in the LLVM issue tracker (https://github.com/llvm/llvm-project/issues), and then follow the procedure described above?

**Reply by HighCommander4** [reply]
Example of this process in action: https://github.com/llvm/llvm-project/issues/56949

**Comment by HighCommander4** [comment]
> **1. Where is the timeline for LLVM point releases documented?** What should I keep an eye on to know that e.g. "the point release LLVM 14.0.y is coming up, and we have until `<this date>` to backport fixes to it"?

This information appears to be announced in "release update" threads on Discourse, such as https://discourse.llvm.org/t/llvm-15-0-x-release-update/65616.

**Comment by sam-mccall** [comment]
Sorry for missing this thread. The process you found is the correct one:

 - we have to file a new bug against llvm/llvm-project, assign the appropriate e.g. 14.0.3 milestone, and write `/cherry-pick <commit-id>` in the bug
 - automation opens a PR, and then pings LLVM owners (myself and kadircet it seems) to approve the cherry pick
 - the milestone has a date on it, that's the deadline for requesting cherry-picks
 - at least in 15.x the releases seem regular and time-based, and then at some point (15.0.3) the release manager said "ok, stable enough" and we don't expect another point release. RM will change next cycle and I suspect the details are personal preference a bit.

**Reply by Yurii-huang** [reply]
I don't know the LSP protocol supported by each version. Obviously, I look for this in the whole network, but I regret that I saw this in the release information of llvm before. It is obvious that clangd has been separated. If there is a separate site recording release notes, including bug repair, new function development, etc., then this is great, and I think it is more meaningful
